### PR TITLE
Update for PRSM

### PR DESCRIPTION
**Pull request by haoxy000**
To support PRSM, we need to update this refresh_bootmap script.

The refreshbootmap on PR/SM will finish:

    find all valid combinations between the compute_fcp and the storage_target_wwpns
    unit_add to loop adding the valid combinations
    mount the volume and update files, executing zipl,etc.
    cleanups: umount the volume, multipath flush, unit_remove, etc.

Please don’t merge this PR. This is just tracking what are changed for PR/SM. I will create another PR to nova-pr/sm repo for these codes. Thanks.

### CODE REVIEW

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 1]
```diff
@@ -408,7 +398,8 @@ function parseArgs {
   isOption -v --verbose "      Print verbose output."      && verbose='-v'
   isOption -x --debug "        Print debugging output."    && debug='-x'
   isOption -t --type "         Device type. The default device typs is fcp and fcp is the only currently supported type."
-  isOption -f --fcpchannel "   FCP channel IDs. Support multi fcpchannel, split by ','."
+  isOption -c --fcpcompute "   Compute FCP channel IDs. Support multi fcpchannel, split by ','."
+  isOption -f --fcpvm "        Target(VM) FCP channel IDs. Support multi fcpchannel, split by ','."
   isOption -w --wwpn "         World Wide Port Name IDs. Support multi wwpn, split by ',' Example: 5005076802400c1b"
   isOption -l --lun "          Logical Unit Number ID. Example: 0000000000000000"
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:404]: Logical Unit Number ID of the volume when mapped to the compute node.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 2]
```diff
@@ -484,7 +479,7 @@ function checkSanity {
   #   Performs basic checks to ensure that a successful deploy can reasonably
   #   be expected.
   if [[ $device_type = 'fcp' ]]; then
-    if [[ ! $lun || ! $fcpchannel ]];then
+    if [[ ! $lun || ! $fcpvm || ! $fcpcompute ]];then
         printError "Exit MSG: Please specify lun and fcpchannel."
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:483]: LUN ID mapped to the compute node, FCP devices on the compute node and FCP devices on the target VM

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 3]
```diff
@@ -408,7 +398,8 @@ function parseArgs {
   isOption -v --verbose "      Print verbose output."      && verbose='-v'
   isOption -x --debug "        Print debugging output."    && debug='-x'
   isOption -t --type "         Device type. The default device typs is fcp and fcp is the only currently supported type."
-  isOption -f --fcpchannel "   FCP channel IDs. Support multi fcpchannel, split by ','."
+  isOption -c --fcpcompute "   Compute FCP channel IDs. Support multi fcpchannel, split by ','."
+  isOption -f --fcpvm "        Target(VM) FCP channel IDs. Support multi fcpchannel, split by ','."
   isOption -w --wwpn "         World Wide Port Name IDs. Support multi wwpn, split by ',' Example: 5005076802400c1b"
   isOption -l --lun "          Logical Unit Number ID. Example: 0000000000000000"
   isOption -d --wwid "         WWID of the volume, Example: 600507640083826de00000000000605b"
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:405]: remove all the following parameters:
1. minfcp
2. ignitionurl
3. nicid
4. ipconfig
5. type

And remove all code related to coreos.

File: zthin-parts/zthin/bin/refresh_bootmap (hunk 2 above)
Comment (dongyanyang) [reply zthin-parts/zthin/bin/refresh_bootmap:483]: remove all other device types and coreos related handlings.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 4]
```diff
@@ -556,25 +551,20 @@ function truncaterdzfcp {
     rd_zfcp=""
     wwpn_idx=0
     selected_paths=0
-    valid_fcps=$(echo ${!valid_paths_dict[*]})
+	# Truncate based on vm fcp
 	while [[ $selected_paths -lt 8 ]]
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:555]: actually we don't know how many to truncate.
This is an open question now, I will open a separate issue to track this.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 5]
```diff
@@ -1028,17 +1022,18 @@ function refreshFCPBootmap {
...
+  inform "RefreshFCPBootmap begins. FCP_VM: ${INPUT_FCPS_VM[*]}, FCP_COMPUTE: ${INPUT_FCPS_COMPUTE[*]}, WWPNs: ${INPUT_WWPNS[*]}, LUN: $lun, WWID: $wwid, minfcp: $minfcp."
 
-  # Check the count of FCP devices, if greater than 8, report error and exit.
-  fcps_len=${#INPUT_FCPS[@]}
+  # Check the count of VM FCP devices, if greater than 8, report error and exit.
+  fcps_len=${#INPUT_FCPS_VM[@]}
   if [ $fcps_len -gt 8 ]
   then
-      printError "Exit MSG: The FCP channel count $fcps_len is greater than 8."
+      printError "Exit MSG: The FCP_VM channel count $fcps_len is greater than 8."
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1036]: actually this is conflict with line 872. : )

But for now, we can keep it as it is.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 6]
```diff
@@ -1091,9 +1070,10 @@ function refreshFCPBootmap {
       fi
   done
 
-  # if the number of online fcp is less than the min FCP number, exit
+  # if the number of online compute fcp is less than the min FCP number, exit
+  onlinefcpscount=${#INPUT_FCPS_COMPUTE[@]}
   if [[ $onlinefcpscount -lt $minfcp ]]; then
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1075]: no need to do this check on minfcp.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 7]
```diff
@@ -1138,25 +1119,26 @@ function refreshFCPBootmap {
...
               ((valid_paths_count=$valid_paths_count+1))
-              # add this (fcp, wwpn) combinations into the rd_zfcp string which will be used to
-              # set the kernel command line in zipl.conf to refresh zipl
-              rd_zfcp+="rd.zfcp=0.0.$fcp,0x$wwpn,$lun "
-          fi
+         fi
       done
   done
 
+  for fcp_vm in "${INPUT_FCPS_VM[@]}"
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1126]: Add a comment before this.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 8]
```diff
@@ -1138,25 +1119,26 @@ function refreshFCPBootmap {
...
+  do
+      rd_zfcp+="rd.zfcp=0.0.$fcp_vm "
+  done
+
   # check whether there is valid combination of FCP and target wwpns found from the `lszfcp -P`
   if [[ $valid_paths_count -eq 0 ]]; then
-      printError "Exit MSG:  No valid path found between FCP devices: ${INPUT_FCPS[*]} and wwpns: ${INPUT_WWPNS[*]}."
+      printError "Exit MSG:  No valid path found between FCP devices: ${INPUT_FCPS_COMPUTE[*]} and wwpns: ${INPUT_WWPNS[*]}."
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1133]: FCP devices: 'xxx' on the compute node and target storage ports: xxx

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 9]
```diff
@@ -1138,25 +1119,26 @@ function refreshFCPBootmap {
...
-  # Usable FCP: Only the FCPs that we can find a valid target WWPN via the `lszfcp -P` can be defined
-  # to VM.
-  # Check whether the count of usable FCPs are bigger than the minimum FCP count
+
+  # Check whether the count of vm FCPs are bigger than the minimum FCP count
   #
-  usable_fcps=(${!valid_paths_dict[*]})
-  if [[ ${#usable_fcps[@]} -lt $minfcp ]]; then
-      printError "Exit MSG: Allocated FCPs include: ${INPUT_FCPS[*]}, but the number of usable FCPs(${usable_fcps[*]}) is less than required minimum FCP count: $minfcp."
+  usable_fcps_count=${#INPUT_FCPS_VM[@]}
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1139]: no need to keep minfcps related changes.

File: zthin-parts/zthin/bin/refresh_bootmap [hunk 10]
```diff
@@ -1207,7 +1189,7 @@ function refreshFCPBootmap {
 
   # check whether there is valid paths found
   if [[ -z $diskPath ]]; then
-    printError "Exit MSG: No avaialble disk found between FCP devices: ${INPUT_FCPS[*]} and wwpns: ${INPUT_WWPNS[*]}"
+    printError "Exit MSG: No avaialble disk found between compute node FCP devices: ${INPUT_FCPS_COMPUTE[*]} and wwpns: ${INPUT_WWPNS[*]}"
```
Comment (dongyanyang) [inline zthin-parts/zthin/bin/refresh_bootmap:1192]: FCP devices xxx on the compute node and the target storage ports: xxx.

File: zthin-parts/zthin/lib/zthinshellutils [hunk 11]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 ###############################################################################
-# Copyright 2017 IBM Corporation
```
Comment (dongyanyang) [inline zthin-parts/zthin/lib/zthinshellutils:3]: just keep 2022

File: zthin-parts/zthin/lib/zthinshellutils [hunk 12]
```diff
@@ -52,11 +52,6 @@ optionHelp='OPTIONS:'
 # becuase it was built into the kernel rather than as a loadable module, so
 # we also (if attempting to load the vmcp module fails) check whether the
 # vmcp command works for us before deciding we have a problem here.
-modprobe vmcp &>/dev/null || vmcp q userid &> /dev/null
```
Comment (dongyanyang) [inline zthin-parts/zthin/lib/zthinshellutils:55]: line 22 - 24 also needs to be removed.

File: zthin-parts/zthin/lib/zthinshellutils [hunk 13]
```diff
@@ -52,11 +52,6 @@ optionHelp='OPTIONS:'
 # becuase it was built into the kernel rather than as a loadable module, so
```
Comment (dongyanyang) [inline zthin-parts/zthin/lib/zthinshellutils:52]: useless comments should be removed.

File: zthin-parts/zthin/lib/zthinshellutils [hunk 14]
```diff
@@ -484,100 +379,8 @@ function chccwdevDevice {
   return $retCode
 } #onOrOffDevice{}
 
-##############################################################################
```
Comment (dongyanyang) [inline zthin-parts/zthin/lib/zthinshellutils:487]: line 267 to line 381 also can be deleted if not used currently.
We only want to keep code that is called.

File: zthin-parts/zthin/lib/zthinshellutils [hunk 15]
```diff
@@ -887,305 +499,9 @@ function removeLun {
   return $rc
 }
 
-###############################################################################
```
Comment (dongyanyang) [inline zthin-parts/zthin/lib/zthinshellutils:890]: function getDiskAlias can also be removed.

**Comment by haoxy000** [comment]
Close this PR. Change to track it in nova-prsm.

**Pull request closed without merging**